week7/answers.sql

```diff
@@ -1,3 +1,270 @@
+  IF effective_damage > 0 THEN
+    SELECT health INTO current_health
+    FROM character_stats
+    WHERE character_id = id_of_character_being_attacked;
+
+    -- Calculate new health of the character
+    SET new_health = current_health - effective_damage;
+
+    -- Update health or delete character if they die
+    IF new_health > 0 THEN
+      UPDATE character_stats
+      SET health = new_health
+      WHERE character_id = id_of_character_being_attacked;
+    ELSE
+    
+      DELETE FROM inventory WHERE character_id = id_of_character_being_attacked;
+      DELETE FROM equipped WHERE character_id = id_of_character_being_attacked;
+      DELETE FROM team_members WHERE character_id = id_of_character_being_attacked;
+      DELETE FROM character_stats WHERE character_id = id_of_character_being_attacked;
+      DELETE FROM characters WHERE character_id = id_of_character_being_attacked;
+    END IF;
+  END IF;
+END$$
+
+CREATE PROCEDURE equip (IN inventory_id INT)
+BEGIN
+    DECLARE char_id INT UNSIGNED;
+    DECLARE item_id INT UNSIGNED;
+
+ SELECT character_id, item_id INTO char_id, item_id
+    FROM inventory
+    WHERE inventory_id = inventory_id;
+-- Move the item from inventory to equipped
+    INSERT INTO equipped (character_id, item_id)
+    SELECT character_id, item_id
+    FROM inventory
+    WHERE inventory_id = inventory_id;
+
+    -- Delete the item from inventory
+    DELETE FROM inventory WHERE inventory_id = inventory_id;
+
+END$$
+
+CREATE PROCEDURE unequip (IN equipped_id INT)
+BEGIN
+   DECLARE char_id INT UNSIGNED;
+    DECLARE item_id INT UNSIGNED;
+-- Get the character ID and item ID associated with the equipped entry
+    SELECT character_id, item_id INTO char_id, item_id
+    FROM equipped
+    WHERE equipped_id = equipped_id;
+
+-- Move the item from equipped to inventory
+    INSERT INTO inventory (character_id, item_id)
+    SELECT character_id, item_id
+    FROM equipped
+    WHERE equipped_id = equipped_id;
+
+    -- Delete the item from equipped
+    DELETE FROM equipped WHERE equipped_id = equipped_id;
+
+END$$
+
+CREATE PROCEDURE set_winners (IN team_id INT)
+BEGIN
+    -- Declare variables for cursor
+    DECLARE done INT DEFAULT 0;
+    DECLARE char_id INT;
+    DECLARE char_name VARCHAR(30);
+
+    -- Declare a cursor for fetching team members
+    DECLARE team_cursor CURSOR FOR
+        SELECT tm.character_id
+        FROM team_members tm
+        JOIN characters_stats cs ON tm.character_id = cs.character_id
+        WHERE tm.team_id = team_id;
+
+    -- Declare a handler for the end of the cursor
+    DECLARE CONTINUE HANDLER FOR NOT FOUND SET done = 1;
+
+    -- Clear the winners table
+    DELETE FROM winners;
+
+    -- Open the cursor
+    OPEN team_cursor;
+
+    -- Loop through each character in the cursor
+winner_loop: LOOP
+    FETCH team_cursor INTO char_id, char_name;
+IF done THEN LEAVE winner_loop;
+END IF;
+
+  INSERT INTO winners (character_id, name)
+    SELECT c.character_id, c.name
```

Comment:
No need for a nested SELECT statement if you use `char_id` and `char_name`. That would be more efficient.